crates/matrix-sdk/src/send_queue/upload.rs

```diff
@@ -542,4 +545,121 @@ impl QueueStorage {
+    pub(super) async fn edit_media_caption(
+        &self,
+        txn: &TransactionId,
+        caption: Option<String>,
+        formatted_caption: Option<FormattedBody>,
+    ) -> Result<Option<AnyMessageLikeEventContent>, RoomSendQueueStorageError> {
+        // This error will be popular here.
+        use RoomSendQueueStorageError::InvalidMediaCaptionEdit;
+
+        let guard = self.store.lock().await;
+        let client = guard.client()?;
+        let store = client.store();
+
+        // The media event can be in one of three states:
+        // - still stored as a dependent request,
+        // - stored as a queued request, active (aka it's being sent).
+        // - stored as a queued request, not active yet (aka it's not being sent yet),
+        //
+        // We'll handle each of these cases one by one.
+
+        {
+            // If the event can be found as a dependent event, update the captions, save it
+            // back into the database, and return early.
+            let dependent_requests = store.load_dependent_queued_requests(&self.room_id).await?;
+
+            if let Some(found) =
+                dependent_requests.into_iter().find(|req| *req.own_transaction_id == *txn)
+            {
+                trace!("found the caption to edit in a dependent request");
+
+                let DependentQueuedRequestKind::FinishUpload {
+                    mut local_echo,
+                    file_upload,
+                    thumbnail_info,
+                } = found.kind
+                else {
+                    return Err(InvalidMediaCaptionEdit);
+                };
+
+                if !update_media_caption(&mut local_echo, caption, formatted_caption) {
+                    return Err(InvalidMediaCaptionEdit);
+                }
+
+                let new_dependent_request = DependentQueuedRequestKind::FinishUpload {
+                    local_echo: local_echo.clone(),
+                    file_upload,
+                    thumbnail_info,
+                };
+                store
+                    .update_dependent_queued_request(
+                        &self.room_id,
+                        &found.own_transaction_id,
+                        new_dependent_request,
+                    )
+                    .await?;
+
+                trace!("caption successfully updated");
+                return Ok(Some(local_echo.into()));
+            }
+        }
+
+        let requests = store.load_send_queue_requests(&self.room_id).await?;
+        let Some(found) = requests.into_iter().find(|req| req.transaction_id == *txn) else {
+            // Couldn't be found anymore, it's not possible to update captions.
+            return Ok(None);
+        };
+
+        trace!("found the caption to edit as a request");
+
+        let QueuedRequestKind::Event { content: serialized_content } = found.kind else {
+            return Err(InvalidMediaCaptionEdit);
+        };
```

Comment:
Just for my understanding: finding a non-event kind here is a developer error and the wrong txn was passed in, right?

Reply:
Indeed, all these errors would either be caused by a wrong txn_id, or a super inconsistent internal state :+1: